pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -583,8 +583,59 @@ void applyGlobalPhase(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
                                                    wires, inverse, params);
 }
 
-template <class PrecisionT, class FuncT> class applyNC2Functor {
+template <class PrecisionT, class FuncT, bool has_controls>
+class applyNC2Functor {};
+
+template <class PrecisionT, class FuncT>
+class applyNC2Functor<PrecisionT, FuncT, true> {
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+
     Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
+    const FuncT core_function;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+
+  public:
+    template <class ExecutionSpace>
+    applyNC2Functor([[maybe_unused]] ExecutionSpace exec,
+                    Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                    std::size_t num_qubits,
+                    const std::vector<std::size_t> &controlled_wires,
+                    const std::vector<bool> &controlled_values,
+                    const std::vector<std::size_t> &wires, FuncT core_function_)
+        : arr(arr_), core_function(core_function_) {
+
+        std::tie(parity, rev_wires) =
```

Comment:
So I tried to use structured binding as suggested in the last PR, however I had the problem that structured binding requires new identifiers. In our case here `parity` and `rev_wires` are both data member of the functor class which are already defined, this needs to be so because `parity` is needed in the `operator()`.

I could use structured binding to some new variables (e.g. `auto& [parity_, rev_wires_] = reverseWires(...)`, then move `parity_` to `parity`, is this solution preferable to `std::tie` @maliasadi @LuisAlfredoNu ?

Reply:
I have now gone with 
```
const auto& [parity_, rev_wires_] =
            reverseWires(num_qubits, wires, controlled_wires);
        parity = parity_;
```
which performs a shallow copy of `parity_` to `parity`